### ISSUE TITLE
Re-import localisations from Crowdin

### ISSUE THREAD

**Issue by barijaona** [ViennaRSS/vienna-rss#933]
Commit 3f4c2e1b0dd67f779fdd8be7df92af5ffadb77fb broke localization of the label attached to the Filter menu.

Running the french localization, it looks like this : 
![menubar_et_vienna__55_non_lus__et_vienna-rss__branch__master_](https://user-images.githubusercontent.com/442030/27982635-7bb5597e-63af-11e7-8380-31ff54e64e7c.png)

Should be "Filtrer par :" instead of "Filter by:"

**Comment by Eitot** [comment]
I still have to re-import all the localisations from Crowdin. With the many changes I plan to do the next weeks, I will do this at the last moment.

**Comment by barijaona** [comment]
Three remarks : 

1. Personally, I'd prefer we re-import the localisations from time to time. It makes things easier for translators in identifying the context.

2. I might be missing something obvious, but I reimported on my local copy the (finished) french translation and the problem remain

3. we'd better check now all texts in Localizable.strings which end with "_title_", "_text_" or "_body_". As I said a few weeks ago, there was a long time ago an attempt to consider them as abstract keys and many of them are meaningless in English. Have a look for instance at "_Empty Trash message text_" or "_Error importing subscriptions body_".

**Comment by Eitot** [comment]
1. It is tedious work, because you have to import each language separately. We'd have to create some helper script, it is still on my list.
2. Have you built the localisations on Crowdin? I see that the last build is from 2 months ago. The XLIFF files have to be compiled from the database.
3. That should indeed happen, but we'd have to change the translation keys manually.

**Comment by barijaona** [comment]
2/ Thanks for the tip. I reimported the french version on my [local copy](https://github.com/barijaona/vienna-rss/tree/localization-test) and the result is a curious mix of french and english. Example : 
![vienna__12_non_lus__et_vienna-rss__branch__localization-test_](https://user-images.githubusercontent.com/442030/27989441-87c35bd4-6441-11e7-9f8e-c46b959eeb6a.png)

3/ This is why I think we should do the corrections now. Otherwise, it would be painful to have to redo it after people have submitted localization updates/fixes.

**Comment by josh64x2** [comment]
@barijaona there are lots of UI changes coming up, it would be a waste of time to reimport them now as they would need to be changed again not long after. These UI changes are more important for the future maintainability of Vienna and so this is a small price to pay especially considering we will not be making a public release from `master` any time soon.  

Let's wait until after the UI changes have been merged into `master` and then make sure we have the translations fixed before we make a release from `master`. We can also setup crowdin or travis CI to automatically build the translations based on any future UI changes.

**Comment by Eitot** [comment]
It appears that these particular strings are not marked as localisable in the source code, neither on master nor on legacy. They simply aren't exported and therefore not on Crowdin either. I cannot say why this worked before, but Crowdin exposed a problem in the source code.

**Comment by barijaona** [comment]
Yes, my point was not to reimport everything, but to test with French if we did break something in between. At a time, we had the [localiseMenus](https://github.com/ViennaRSS/vienna-rss/blob/legacy3.1.x/src/AppController.m#L269-L296) method, but we switched philosophy in between…

**Comment by Eitot** [comment]
Addendum: I think the problem is this.

The mentioned strings are created [here](https://github.com/ViennaRSS/vienna-rss/blob/master/src/ArticleFilter.m#L101-L107) (note the lack of `NSLocalizedString`): 
```objc
[ArticleFilter createFilter:@"Unread or flagged" tag:MA_Filter_Unread_Or_Flagged comparator:@selector(unreadOrFlaggedArticleFilterComparator:)];
```

Now, these strings are used [here](https://github.com/ViennaRSS/vienna-rss/blob/master/src/AppController.m#L1645-L1660), but wrapped in `NSLocalizedString`:
```objc
NSMenuItem * menuItem = [[NSMenuItem alloc] initWithTitle:NSLocalizedString([filter name], nil) action:@selector(changeFiltering:) keyEquivalent:@""];
```

This is a misuse of `NSLocalizedString` (it is a preprocessor macro). What happens is that at runtime, the localised strings are indeed replaced, but only if the keys are translated. However, because the strings are assigned at runtime, Xcode will neither export nor import them. Unused keys are discarded upon importing a new XLIFF file, so this is why the translations are gone.

**Comment by barijaona** [comment]
Another method worth mentioning is `-localiseHeaderStrings` in TableViewExtensions

**Comment by Eitot** [comment]
There are more:
- AppController.m
- Database.m
- NewSubscription.m
- PluginManager.m
- SyncingPreferencesViewController.m

**Comment by barijaona** [comment]
Yeah… looks like, at the times of Vienna 2.0, the philosophy was notably different from today's dominant one : much more UI was created programmatically, and the dynamic nature of Objective-C was used for cool tricks 😸

**Comment by Eitot** [comment]
Anyone picking this up? I still have to wrap up #930 and have already a follow-up on that one.

**Comment by barijaona** [comment]
I’ll have a look at it.
I would not be too assertive regarding timeline : let’s have a public beta first, and wait feedback…

**Comment by Eitot** [comment]
I have uploaded the source strings and existing translations again on Crowdin. Once the changes are made in code, I can export the hopefully complete translation files.

**Comment by barijaona** [comment]
Seems not to be fixed by reimport of translation files after #940.

**Comment by Eitot** [comment]
It wasn't supposed to fix anything, just to make sure that the strings database matches master. We still need to get rid of the dynamic NSLocalizedString before Xcode exports/imports the strings.

**Comment by barijaona** [comment]
The problem is not related to Vienna code, but to Crowdin import and export.
If I use a text editor to modify the `fr.xliff` file exported by Xcode then directly reimport it, the updated translations load correctly.

With the `Vienna RSS.xliff` file downloaded from Crowdin, I need to hack the `<file original=` and `<trans-unit id=`> attributes to be able to reimport correctly the updated translations in Xcode.

**Comment by Eitot** [comment]
Give me a concrete example, please. I do not see what you mean.

If Vienna would correctly identify all the localisable strings upon export, the XLIFF file uploaded to Crowdin would be complete and have the correct IDs to begin with. Manual manipulation of the XLIFF file is what we want to avoid.

There is no way around rewriting the code to make sure that the XLIFF tool of Xcode works as intended. NSLocalizedString is meant to be used with static strings, per Apple's instructions.

**Comment by barijaona** [comment]
Sorry. I was confused by the download possibilities at https://crowdin.com/project/vienna-rss/fr#
There is « Download » and « Download in XLIFF » and their effects are completely different 😆 

Currently, my problem is limited to : 
- the headings in the article list
- and the menus in the dialog sheet when you create/edit a smart folder
They require a few more items to be present in Localizable.strings
I see some workarounds, but I am still searching for an elegant one.

**Comment by josh64x2** [comment]
@barijaona I was going to move the smart folder options translations into the xib, and assign a tag to each option which I can use (maybe as an enum) in the code itself.

**Comment by barijaona** [comment]
@josh64x2 OK, try going ahead with this.

**Comment by Eitot** [comment]
I went ahead and fixed most of the other NSLocalizedString issues. Just two areas remain: the mentioned smart-folder keys (SearchFolder and Criteria classes) and the table headers (in Database and TableViewExtensions). The table headers are based on the constants defined in Article.m.

**Comment by barijaona** [comment]
@Eitot I hope my latest pull request does not duplicate any efforts you made in between…

**Comment by Eitot** [comment]
Nope, seems to match perfectly. Parallel work!
